text/0000-feature-metadata.md

```diff
@@ -0,0 +1,309 @@
+-   `unstable` and `public` uses may not be common enough to be worth including
+-   A markdown parser will eventaully be required to properly parse the `doc`
+    field. Presenting unrendered markdown is likely suitable as a first step,
+    but in general, markdown is required to fully interpret the field.
+-   This RFC does not provide any way for `rustdoc` to get the information it
+    requires. This will require separate design work.
+-   There is no way to structure features in a way that they are split into
+    sections, unlike with the `document-features` crate.
+-   Features cannot be ordered since the TOML specification does not allow it.
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+-   Feature descriptions could be specified somewhere in Rust source files. This
+    has the downside of creating multiple sources of truth on features.
+-   Cargo could parse doc comments in `Cargo.toml`, like the `document-features`
+    crate (linked below).
+
+    ```toml
+    [features]
+    foo = { requires = [], doc = "foo feature" }
+    ## foo feature
+    foo = []
+    ```
+
+    This was decided against as part of this RFC because it would mean that
+    standard TOML parsers (including anything `serde`-based) cannot extract all
+    the information in `Cargo.toml` - a format-preserving parser would be
+    required. This differs from documentation in Rust source, where a custom
+    parser already has to be used.
+
+# Prior art
+
+[prior-art]: #prior-art
+
+-   There is an existing crate that uses TOML comments to create a features table:
+    <https://docs.rs/document-features/latest/document_features/>
+-   `docs.rs` displays a feature table, but it is fairly limited. If features
+    start with `_`, they are hidden from this table.
+-   Ivy has a [visibility attribute] for its configuration (mentioned in [cargo #10882])
+
+# Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+-   Do we want all the proposed keys? Specifically, `public` and `unstable` may be
+    more than what is needed.
+
+    See also:
+
+    -   <https://github.com/rust-lang/cargo/issues/10882>
+    -   <https://github.com/rust-lang/cargo/issues/10881>
+
+-   If we use the semantics as-written, should there be a
+    `--allow-private-features` or `--allow-unstable-features` flag? Or how should
+    a user opt in?
+-   Does it make sense to have separate `hidden` (not documented) and `public`
+    (feature not allowed downstream) attribute? I think probably not
+-   Should there be a way to deny deprecated features?
+-   The Cargo index may need a way to be aware of deprecated features, so it can
+    properly report them during resolution. What would be needed here?
+
+It is worth noting that simpler keys (`requires`, `doc`, `deprecated`) could be
+stabilized immediately and other features could be postponed.
+
+# Future possibilities
```

Comment:
Private and/or unstable features might be sufficient to unblock rust-lang/cargo#1596.

Reply:
Awesome thought! I added it under future possibiliites here fff9af1744999c5409e55e23873d54ff19023f2a.

Do you have any ideas about the semantics of this might work? I'll add some details, which would be useful if we wind up splitting the RFCs.

Reply:
I did see your comment - I was mostly wondering about whether setting `public=false` on a dependency feature flag would be sufficient, or if you were imagining a separate change that would follow this RFC

Reply:
There might be additional work after this RFC.  We'd need to discuss in that issue how exactly we would want to build on this.

Reply:
Thanks for confirming, I will mark this as resolved then